File: backend/src/main/java/com/slpolice/trafficfines/payment/dto/PaymentDTO.java
package com.slpolice.trafficfines.payment.dto;

import com.slpolice.trafficfines.payment.entity.Payment;
import com.slpolice.trafficfines.payment.entity.PaymentChannel;
import com.slpolice.trafficfines.payment.entity.PaymentMethod;
import lombok.Builder;
import lombok.Data;

import java.math.BigDecimal;
import java.time.LocalDateTime;

/**
 * Data Transfer Object for a Payment — used in all API responses.
 * As per the DTO Pattern: "Never expose JPA entities directly."
 *
 * This is the agreed API contract shape for the payment receipt,
 * shared with the frontend teams (Member 3, 4).
 */
@Data
@Builder
public class PaymentDTO {

    private Long id;

    // Associated fine info (for receipt display)
    private Long fineId;
    private String referenceNumber;
    private String driverName;
    private String vehicleNumber;
    private String categoryDescription;

    // Payment details
    private BigDecimal amountPaid;
    private PaymentMethod paymentMethod;
    private PaymentChannel paymentChannel;
    private String transactionRef;
    private LocalDateTime paidAt;

    /**
     * Static factory method to convert a Payment entity to a PaymentDTO.
     */
    public static PaymentDTO from(Payment payment) {
        return PaymentDTO.builder()
                .id(payment.getId())
                .fineId(payment.getFine().getId())
                .referenceNumber(payment.getFine().getReferenceNumber())
                .driverName(payment.getFine().getDriverName())
                .vehicleNumber(payment.getFine().getVehicleNumber())
                .categoryDescription(payment.getFine().getCategory().getDescription())
                .amountPaid(payment.getAmountPaid())
                .paymentMethod(payment.getPaymentMethod())
                .paymentChannel(payment.getPaymentChannel())
                .transactionRef(payment.getTransactionRef())
                .paidAt(payment.getPaidAt())
                .build();
    }
}

File: mobile/lib/data/models/payment_dto.dart
class PaymentDTO {
  final int id;
  final int fineId;
  final String referenceNumber;
  final String driverName;
  final String vehicleNumber;
  final String categoryDescription;
  final double amountPaid;
  final String paymentMethod;
  final String paymentChannel;
  final String transactionRef;
  final String paidAt;

  PaymentDTO({
    required this.id,
    required this.fineId,
    required this.referenceNumber,
    required this.driverName,
    required this.vehicleNumber,
    required this.categoryDescription,
    required this.amountPaid,
    required this.paymentMethod,
    required this.paymentChannel,
    required this.transactionRef,
    required this.paidAt,
  });

  factory PaymentDTO.fromJson(Map<String, dynamic> json) {
    return PaymentDTO(
      id: json['id'] ?? 0,
      fineId: json['fineId'] ?? 0,
      referenceNumber: json['referenceNumber'] ?? '',
      driverName: json['driverName'] ?? '',
      vehicleNumber: json['vehicleNumber'] ?? '',
      categoryDescription: json['categoryDescription'] ?? '',
      amountPaid: (json['amountPaid'] ?? 0).toDouble(),
      paymentMethod: json['paymentMethod'] ?? '',
      paymentChannel: json['paymentChannel'] ?? '',
      transactionRef: json['transactionRef'] ?? '',
      paidAt: json['paidAt'] ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'fineId': fineId,
      'referenceNumber': referenceNumber,
      'driverName': driverName,
      'vehicleNumber': vehicleNumber,
      'categoryDescription': categoryDescription,
      'amountPaid': amountPaid,
      'paymentMethod': paymentMethod,
      'paymentChannel': paymentChannel,
      'transactionRef': transactionRef,
      'paidAt': paidAt,
    };
  }
}
